go/extractor/cli/go-autobuilder/go-autobuilder.go

```diff
@@ -727,62 +727,54 @@ func outsideSupportedRange(version string) bool {
-// called, so any versions that are found are within the supported range.
-func checkForVersionsNotFound(v versionInfo) (msg, version string) {
-	if !v.goEnvVersionFound && !v.goModVersionFound {
+// Assuming `v.goModVersionFound` is false, emit a diagnostic and return the version to install,
+// or the empty string if we should not attempt to install a version of Go.
+func checkForGoModVersionNotFound(v versionInfo) (msg, version string) {
+	if !v.goEnvVersionFound {
 		msg = "No version of Go installed and no `go.mod` file found. Writing an environment " +
 			"file specifying the maximum supported version of Go (" + maxGoVersion + ")."
 		version = maxGoVersion
 		diagnostics.EmitNoGoModAndNoGoEnv(msg)
-	}
-
-	if !v.goEnvVersionFound && v.goModVersionFound {
-		msg = "No version of Go installed. Writing an environment file specifying the version " +
-			"of Go found in the `go.mod` file (" + v.goModVersion + ")."
-		version = v.goModVersion
-		diagnostics.EmitNoGoEnv(msg)
-	}
-
-	if v.goEnvVersionFound && !v.goModVersionFound {
+	} else if outsideSupportedRange(v.goEnvVersion) {
+		msg = "No `go.mod` file found. The version of Go installed in the environment (" +
+			v.goEnvVersion + ") is outside of the supported range (" + minGoVersion + "-" +
+			maxGoVersion + "). Writing an environment file specifying the maximum supported " +
+			"version of Go (" + maxGoVersion + ")."
+		version = maxGoVersion
+		diagnostics.EmitNoGoModAndGoEnvUnsupported(msg)
+	} else {
 		msg = "No `go.mod` file found. Version " + v.goEnvVersion + " installed in the " +
-			"environment. Writing an environment file not specifying any version of Go."
+			"environment is supported. Writing an environment file not specifying any " +
+			"version of Go."
 		version = ""
-		diagnostics.EmitNoGoMod(msg)
+		diagnostics.EmitNoGoModAndGoEnvSupported(msg)
 	}
 
 	return msg, version
 }
 
-// Compare `v.goModVersion` and `v.goEnvVersion`. emit a diagnostic and return the version to
-// install, or the empty string if we should not attempt to install a version of Go. We assume that
-// `checkForUnsupportedVersions` and `checkForVersionsNotFound` have already been called, so both
-// versions are found and are within the supported range.
-func compareVersions(v versionInfo) (msg, version string) {
-	if semver.Compare("v"+v.goModVersion, "v"+v.goEnvVersion) > 0 {
+// Assuming `v.goModVersionFound` is true, emit a diagnostic and return the version to install,
+// or the empty string if we should not attempt to install a version of Go.
+func checkForGoModVersionFound(v versionInfo) (msg, version string) {
+	if outsideSupportedRange(v.goModVersion) {
+		msg = "The version of Go found in the `go.mod` file (" + v.goModVersion +
+			") is outside of the supported range (" + minGoVersion + "-" + maxGoVersion +
+			"). Writing an environment file not specifying any version of Go."
+		version = ""
+		diagnostics.EmitUnsupportedVersionGoMod(msg)
```

Comment:
Minor niggle, but it might be nice to capture some of the rationale behind the logic that is implemented here in comments for each condition/outcome. I know we have your EDR and the Slack thread in which these decisions are discussed, but they will eventually disappear / be forgotten about, while adding the rationale here will persist.